provd/internal/services/gdm/gdm_test.go

```diff
@@ -0,0 +1,172 @@
+package gdm_test
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/canonical/ubuntu-desktop-provision/provd/internal/services/gdm"
+	"github.com/canonical/ubuntu-desktop-provision/provd/internal/testutils"
+	pb "github.com/canonical/ubuntu-desktop-provision/provd/protos"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		invalidGdmPrefix bool
+
+		wantErr bool
+	}{
+		// Success case
+		"Success on creating GDM service": {},
+
+		// Error case
+		"Error on session bus unavailable": {invalidGdmPrefix: true, wantErr: true},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			var opts []gdm.Option
+			if tc.invalidGdmPrefix {
+				opts = []gdm.Option{
+					gdm.WithGdmPrefix("invalid"),
+				}
+			}
+
+			_, err := gdm.New(testutils.NewDbusConn(t), opts...)
+			if tc.wantErr {
+				require.Error(t, err, "Expected error, but got nil")
+				return
+			}
+			require.NoError(t, err, "Expected no error, but got %v", err)
+		})
+	}
+}
+
+func TestLaunchDesktopSession(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username  string
+		password  string
+		cancelReq bool
+
+		wantErr bool
```

Comment:
2: ...You could then have flags for cases where your side effects fail that you pass to the mocks